pkg/localstorev2/internal/cache/cache.go

```diff
@@ -0,0 +1,379 @@
+	if addr.IsZero() {
+		return make([]byte, swarm.HashSize)
+	}
+	return addr.Bytes()
+}
+
+type cacheEntry struct {
+	Address swarm.Address
+	Prev    swarm.Address
+	Next    swarm.Address
+}
+
+func (cacheEntry) Namespace() string { return "entry" }
+
+func (c *cacheEntry) ID() string { return c.Address.ByteString() }
+
+func (c *cacheEntry) Marshal() ([]byte, error) {
+	entryBuf := make([]byte, cacheEntrySize)
+	if c.Address.IsZero() {
+		return nil, errMarshalCacheEntryInvalidAddress
+	}
+	copy(entryBuf[:swarm.HashSize], c.Address.Bytes())
+	copy(entryBuf[swarm.HashSize:2*swarm.HashSize], addressBytesOrZero(c.Prev))
+	copy(entryBuf[2*swarm.HashSize:], addressBytesOrZero(c.Next))
+	return entryBuf, nil
+}
+
+func (c *cacheEntry) Unmarshal(buf []byte) error {
+	if len(buf) != cacheEntrySize {
+		return errUnmarshalCacheEntryInvalidSize
+	}
+	newEntry := new(cacheEntry)
+	newEntry.Address = swarm.NewAddress(append(make([]byte, 0, swarm.HashSize), buf[:swarm.HashSize]...))
+	newEntry.Prev = getAddressOrZero(buf[swarm.HashSize : 2*swarm.HashSize])
+	newEntry.Next = getAddressOrZero(buf[2*swarm.HashSize:])
+	*c = *newEntry
+
+	return nil
+}
+
+const cacheStateSize = 2*swarm.HashSize + 8
+
+type cacheState struct {
+	Start swarm.Address
+	End   swarm.Address
+	Count uint64
+}
+
+func (cacheState) Namespace() string { return "state" }
+
+func (cacheState) ID() string { return "e" }
+
+func (c *cacheState) Marshal() ([]byte, error) {
+	entryBuf := make([]byte, cacheStateSize)
+	copy(entryBuf[:swarm.HashSize], addressBytesOrZero(c.Start))
+	copy(entryBuf[swarm.HashSize:2*swarm.HashSize], addressBytesOrZero(c.End))
+	binary.LittleEndian.PutUint64(entryBuf[2*swarm.HashSize:], c.Count)
+	return entryBuf, nil
+}
+
+func (c *cacheState) Unmarshal(buf []byte) error {
+	if len(buf) != cacheStateSize {
+		return errUnmarshalCacheStateInvalidSize
+	}
+	newEntry := new(cacheState)
+	newEntry.Start = getAddressOrZero(buf[:swarm.HashSize])
+	newEntry.End = getAddressOrZero(buf[swarm.HashSize : 2*swarm.HashSize])
+	newEntry.Count = binary.LittleEndian.Uint64(buf[2*swarm.HashSize:])
+	*c = *newEntry
+	return nil
+}
+
+// Cache is the part of the localstore which keeps track of the chunks that are not
+// part of the reserve but are potentially useful to store for obtaining bandwidth
+// incentives. In order to avoid GC we will only keep track of a fixed no. of chunks
+// as part of the cache and evict a chunk as soon as we go above capacity. In order
+// to achieve this we will use some additional cache state in-mem and stored on disk
+// to create a double-ended queue. The typical operations required here would be
+// a pushBack which adds an item to the end and popFront, which removed the first item.
+// The different operations:
+// 1. New item will be added to the end.
+// 2. Item pushed to end on access.
+// 3. Removal happens from the front.
+type Cache struct {
+	mtx      sync.Mutex
+	state    *cacheState
+	capacity uint64
```

Comment:
So not clear on why we need this? Can you explain a bit more.

The reason for keeping it uint is that it will be always greater than 0 and uint64 as this is way more than any realistic max value the user can set. So it shouldnt need to change.

Reply:
In Golang, it is generally recommended to use the `int` type if you know you won't need the extra range of the `int64` type, even if you know the value won't be negative (look at the `cap` and `length` slice fields). Also, as I mentioned earlier, using `int` eliminates unnecessary casting and bugs such as:
`for i := 0; i < c.capacity - 3; i++`

Reply:
Will do this in a cleanup PR that I have planned after all the transactions and localstore container is done.